src/test/java/com/example/onlinebookstore/controller/BookControllerTest.java

```diff
@@ -0,0 +1,126 @@
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.http.MediaType;
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+
+@ExtendWith(MockitoExtension.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class BookControllerTest {
+
+    @Autowired
+    private static MockMvc mockMvc;
+
+    @Mock
+    private BookService bookService;
+
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    @InjectMocks
+    private BookController bookController;
+
+    private BookRequestDto bookRequestDto;
+
+    private BookResponseDto bookResponseDto;
+
+    @BeforeAll
+    static void beforeAll(@Autowired WebApplicationContext applicationContext) {
+        mockMvc = MockMvcBuilders
+                .webAppContextSetup(applicationContext)
+                .apply(springSecurity())
+                .build();
+    }
+
+    @BeforeEach
+    void setUp() {
+        bookRequestDto = new BookRequestDto();
+        bookRequestDto.setTitle("Sample Book 3");
+        bookRequestDto.setIsbn("9781122334457");
+        bookRequestDto.setPrice(BigDecimal.valueOf(29.99));
+        bookRequestDto.setAuthor("Author C");
+        bookRequestDto.setDescription("Yet another sample book description.");
+        bookRequestDto.setCategoryIds(Set.of(1L, 2L));
+        bookRequestDto.setCoverImage("http://example.com/cover3.jpg");
+
+        bookResponseDto = new BookResponseDto();
+        bookResponseDto.setId(1L);
+        bookResponseDto.setTitle(bookRequestDto.getTitle());
+        bookResponseDto.setAuthor(bookRequestDto.getAuthor());
+        bookResponseDto.setPrice(bookRequestDto.getPrice());
+        bookResponseDto.setIsbn(bookRequestDto.getIsbn());
+        bookResponseDto.setDescription(bookRequestDto.getDescription());
+        bookResponseDto.setCoverImage(bookRequestDto.getCoverImage());
+        bookResponseDto.setCategoryIds(bookRequestDto.getCategoryIds());
+    }
+
+    @WithMockUser(username = "user", roles = {"ADMIN"})
+    @Test
+    public void testGetAllBooks() throws Exception {
+        // When
+        String result = mockMvc.perform(get("/books").contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn()
+                .getResponse()
+                .getContentAsString();
+
+        // Then
+        List<BookResponseDto> actual = objectMapper.readValue(result,
+                new TypeReference<>() {
+                });
+
+        // Verify
+        assertNotNull(actual);
+        assertEquals(3, actual.size());
```

Comment:
we can also check fields that it returns here and not just the size, same for other tests